src/main/java/com/hierynomus/msfscc/fileinformation/ShareInfo.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (C)2016 - SMBJ Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hierynomus.msfscc.fileinformation;
+
+import java.math.BigInteger;
+
+import com.hierynomus.protocol.commons.buffer.Buffer;
+import com.hierynomus.protocol.commons.buffer.Buffer.BufferException;
+
+public class ShareInfo {
+
+	private final long totalAllocationUnits;
+	private final long callerAvailableAllocationUnits;
+	private final long actualAvailableAllocationUnits;
+	private final int sectorsPerAllocationUnit;
+	private final int bytesPerSector;
+	
+	private final long totalSpace;
+	private final long callerFreeSpace;
+	private final long actualFreeSpace;
+	
+	private final static BigInteger MAX_VALUE = BigInteger.valueOf(Long.MAX_VALUE);
+	
+	ShareInfo(long totalAllocationUnits, long callerAvailableAllocationUnits,
+			long actualAvailableAllocationUnits, int sectorsPerAllocationUnit, int bytesPerSector) {
+		this.totalAllocationUnits = totalAllocationUnits;
+		this.callerAvailableAllocationUnits = callerAvailableAllocationUnits;
+		this.actualAvailableAllocationUnits = actualAvailableAllocationUnits;
+		this.sectorsPerAllocationUnit = sectorsPerAllocationUnit;
+		this.bytesPerSector = bytesPerSector;
+
+		// Using BigInteger to check for overflows...
+		BigInteger bytesPerAllocationUnit = BigInteger.valueOf(sectorsPerAllocationUnit).multiply(BigInteger.valueOf(bytesPerSector));
+		assert bytesPerAllocationUnit.compareTo(MAX_VALUE) <= 0;

```

Comment:
Assertions are not turned on in most systems. Please use runtime checks if these are important.